Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -105,154 +105,289 @@ struct TextBox::TextValidator : QValidator {
+      auto& current_highlight = m_highlight->get();
+      setCursorPosition(current_highlight.m_end);
+      setSelection(current_highlight.m_start, get_size(current_highlight));
+      setValidator(m_text_validator);
+      installEventFilter(this);
+      connect(this, &QLineEdit::editingFinished, this,
+        &LineEdit::on_editing_finished);
+      connect(
+        this, &QLineEdit::textEdited, this, &LineEdit::on_text_edited);
+      m_current_connection = m_current->connect_update_signal(
+        [=] (const auto& value) { on_current(value); });
+      connect(this, &QLineEdit::cursorPositionChanged, this,
+        std::bind_front(&LineEdit::on_cursor_position, this));
+      connect(this, &QLineEdit::selectionChanged, this,
+        std::bind_front(&LineEdit::on_selection, this));
+      m_highlight_connection = m_highlight->connect_update_signal(
+        std::bind_front(&LineEdit::on_highlight, this));
+      m_text_box->setCursor(cursor());
+      m_text_box->setFocusPolicy(focusPolicy());
+      m_text_box->setFocusProxy(this);
+      m_text_box->installEventFilter(this);
+      auto layout = new QHBoxLayout(m_text_box);
+      layout->setContentsMargins(0, 0, 0, 0);
+      layout->addWidget(this);
+      add_pseudo_element(*m_text_box, Placeholder());
+      m_placeholder_style_connection = connect_style_signal(
+        *m_text_box, Placeholder(), [=] { on_placeholder_style(); });
+      m_read_only_connection = read_only->connect_update_signal(
+        [=] (auto value) { on_read_only(value); });
+      setReadOnly(read_only->get());
+    }
+
+    void set_display_text(const QString& text) {
+      m_text_validator->m_is_text_elided = text != m_current->get();
+      setText(text);
+      if(m_text_validator->m_is_text_elided) {
+        setCursorPosition(0);
+      }
     }
 
     void set_placeholder(const QString& placeholder) {
       m_placeholder = placeholder;
-      update_elision();
+      elide_placeholder_text();
       update();
     }
 
-    void set_placeholder_visible(bool is_visible) {
-      if(is_visible != m_is_placeholder_visible) {
-        m_is_placeholder_visible = is_visible;
-        update();
+    void set_style(
+        const BoxGeometry& geometry, const TextStyleProperties& text_style) {
+      m_text_box->layout()->setContentsMargins(
+        geometry.get_padding_left(), geometry.get_padding_top(),
+        geometry.get_padding_right(), geometry.get_padding_bottom());
+      auto stylesheet = QString(
+        R"(#0x%1 {
+          background: transparent;
+          border-width: 0px;
+          padding: 0px;)").arg(reinterpret_cast<std::intptr_t>(this));
+      auto map = StyleSheetMap([] {});
+      map.set("color", text_style.m_text_color);
+      map.write(stylesheet);
+      if(text_style.m_alignment != alignment()) {
+        setAlignment(text_style.m_alignment);
+      }
+      auto font = text_style.m_font;
+      font.setPixelSize(text_style.m_size);
```

Comment:
Can you set the pixel size directly in ```text_style.m_font``` instead of having a separate field ```text_style.m_size``` for it?